### PR TITLE
Initial Plugin

### PR DESCRIPTION
**Pull request by byronic**
This PR contains the initial work on the (eventual) 0.1.0 release of this plugin.

As of the time of this writing, executing this plugin during a mvn site run correctly generates a table of contents + report set on the Audits page it adds to the maven site.

### CODE REVIEW

File: examples/example-reporter-impl/pom.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.infrastructurebuilder</groupId>
+    <artifactId>ibparent</artifactId>
+    <version>24</version>
```
Comment (mykelalvis) [inline examples/example-reporter-impl/pom.xml]: 28

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.infrastructurebuilder</groupId>
+    <artifactId>ibparent</artifactId>
+    <version>26</version>
```
Comment (mykelalvis) [inline pom.xml]: 28

File: src/main/java/org/infrastructurebuilder/auditor/AuditorReportMojo.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.maven.reporting.AbstractMavenReport;
+import org.apache.maven.reporting.MavenReportException;
+
+@Mojo(name = "audit", defaultPhase = LifecyclePhase.SITE, requiresDependencyResolution = ResolutionScope.RUNTIME, requiresProject = true, threadSafe = true)
+public class AuditorReportMojo extends AbstractMavenReport {
+
+  /**
+   * Where the HTML pages of the report will be created
+   */
+  @Parameter(defaultValue = "${project.reporting.outputDirectory}", property = "outputDirectory", required = true)
```
Comment (mykelalvis) [inline src/main/java/org/infrastructurebuilder/auditor/AuditorReportMojo.java]: prolly `property = "audit.outputDirectory"`

File: src/main/java/org/infrastructurebuilder/auditor/AuditorReportMojo.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+   * Practical reference to the maven project
+   */
+  @Parameter(defaultValue = "${project}", readonly = true)
+  private MavenProject project;
+
+  @Component
+  private PageGenerator pageGenerator;
+
+  public void setPageGenerator(PageGenerator pageGenerator) {
+    this.pageGenerator = Objects.requireNonNull(pageGenerator);
```
Comment (mykelalvis) [inline src/main/java/org/infrastructurebuilder/auditor/AuditorReportMojo.java]: `import static java.util.Objects.requireNonNull`

File: src/main/java/org/infrastructurebuilder/auditor/PageGenerator.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    p.parse(confidentiality, sink);
+    sink.paragraph_();
+
+    sink.paragraph();
+    p.parse(intro, sink);
+    sink.paragraph_();
+
+    long duration = Duration.between(results.getTimestampStart().toInstant(), results.getTimestampEnd().toInstant())
+        .toMillis();
+    resultMetrics(results.getDescriptionHeaders(), results.getResults(), duration, sink);
```
Comment (mykelalvis) [inline src/main/java/org/infrastructurebuilder/auditor/PageGenerator.java:118]: `Duration` is an available type.  Mebbe `AuditResults` has a `Duration getDuration()`?

File: src/test/java/org/infrastructurebuilder/auditor/TestAuditorReportMojo.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class TestAuditorReportMojo {
+  @Test
+  public void testConstructor() {
+    AuditorReportMojo m = new AuditorReportMojo();
+    assertEquals("audits", m.getOutputName());
```
Comment (mykelalvis) [inline src/test/java/org/infrastructurebuilder/auditor/TestAuditorReportMojo.java]: Set the setters, yo.

File: examples/example-audit-execution/pom.xml [hunk 7]
```diff
@@ -0,0 +1,71 @@
+<!--
```
Comment (mykelalvis) [inline examples/example-audit-execution/pom.xml]: This should probably be in `src/it/example` and executed by invoker (somewhere in the tree).

**Review by mykelalvis — CHANGES_REQUESTED** [review]

File: src/main/java/org/infrastructurebuilder/auditor/PageGenerator.java (hunk 5 above)
Comment (byronic) [reply src/main/java/org/infrastructurebuilder/auditor/PageGenerator.java:118]: Counterpoint:

```
Caused by: org.codehaus.modello.model.ModelValidationException: Field 'duration': Illegal type: 'java.time.Duration'.
    at org.codehaus.modello.model.ModelField.validateElement (ModelField.java:224)
```

Comment (mykelalvis) [reply src/main/java/org/infrastructurebuilder/auditor/PageGenerator.java:118]: Sorry.  `Duration` is an available *Java* type.

**Review by mykelalvis — APPROVED** [review]

File: pom.xml [hunk 8]
```diff
@@ -20,23 +20,35 @@
...
-    <test.coverage.percentage.required>28</test.coverage.percentage.required>
+    <test.coverage.percentage.required>78</test.coverage.percentage.required>
   </properties>
 
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>org.infrastructurebuilder.util</groupId>
+        <artifactId>ibcore-bom</artifactId>
+        <version>0.15.0.0</version>
```
Comment (mykelalvis) [inline pom.xml]: 0.16.0.1 is out....

**Review by mykelalvis — CHANGES_REQUESTED** [review]
No requirement for moving to the new core, but recommended.

**Comment by mykelalvis** [comment]
The build fails.  Looks like the invoker is being invoked before install happens?  Maybe the invoker plugin got an upgrade, because I see some new goals available in it and I didn't really keep up with it for the last few months.  Let's take a look at the goals other than `invoker:run` .

**Comment by mykelalvis** [comment]
I added `invoker:install` before `invoker:run` as a test and it gets farther, but you can't have one integration test depend on another unless you produce ordering of the names themselves.  (`example-audit-execution` happens before it's dependency `example-reporter-impl` is built and installed).

I created #2 to denote the changes that fix this.

**Comment by byronic** [comment]
I'll bet I was getting a false positive from the local repository level re: dependencies. Let me clean local and rebuild, see if anything else shakes out

**Comment by mykelalvis** [comment]
Ship it

**Pull request merged**